### PR TITLE
[13.0] Create module account_financial_discount

### PR DESCRIPTION
**Pull request by grindtildeath**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
This already exists: https://github.com/OCA/account-invoicing/tree/12.0/account_global_discount

**Comment by grindtildeath** [comment]
@pedrobaeza 

1. this is still a draft
2. the module you mention doesn't provide the same functionalities ;)

**Comment by pedrobaeza** [comment]
Well, it's more or less the same, but yes, it's not the same, but @sbidoul can say about an exact replacement of this module that they develop.

**Comment by grindtildeath** [comment]
@pedrobaeza If his module hasn't been submitted in the OCA yet, what's the point of recommending it here? Sorry but I don't understand your logic here...
Anyway you're still welcome to review and propose improvements on this PR :)

**Comment by pedrobaeza** [comment]
Well, I don't look for anything outside OCA, but my comment was from memory for avoiding your extra work if exists. I have looked, and the module I was mentioning is: https://odoo-community.org/shop/product/account-cash-discount-base-4612

You also have from your company https://github.com/OCA/account-payment/pull/231

**Comment by sbidoul** [comment]
Yes, there is this suite of cash discount modules in v10. It works well and can handle different VAT scenarios.
I *think* we'll port them to v13. I'll check with my colleagues.

**Comment by grindtildeath** [comment]
Well I just gave a look to account_cash_discount_base and depending modules in account-payment v10 and must say it is close to what we did here.
But with the all the changes between 10.0 and 13.0 on the reconciliation part, IMO we should try to merge these modules together, if not for v13.0 at least for v14.0.

But  account_early_payment_discount is also missing a lot of features on this and that's why we developed this new module.

fyi @fclementic2c

**Comment by sbidoul** [comment]
cc/ @Cedric-Pigeon @benwillig @chrisdec

**Comment by sbidoul** [comment]
cross-referencing the PR for the cash_discount modules v13: https://github.com/OCA/account-payment/pull/334

File: account_financial_discount/models/account_payment_term.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                        ('state', '=', 'posted'),
+                    ]
+                )
+                if rec_count > 0:
+                    raise UserError(
+                        _(
+                            'You cannot change the discount on payment term %s since some posted journal entries already exist'
+                        )
+                        % term.name
+                    )
```
Comment (santostelmo) [inline account_financial_discount/models/account_payment_term.py:48]:
```suggestion
                    )
        return super().write(values)
```

Comment (grindtildeath) [reply account_financial_discount/models/account_payment_term.py:48]: :scream:

**Comment by pedrobaeza** [comment]
AFAIK, Acsone was doing something very similar. @sbidoul can you post the link to your work?

**Comment by sbidoul** [comment]
Yes, they have been merged in OCA/account-payment: https://github.com/OCA/account-payment/tree/14.0/account_cash_discount_base and https://github.com/OCA/account-payment/tree/14.0/account_cash_discount_payment

**Comment by grindtildeath** [comment]
@pedrobaeza @sbidoul We considered account_cash_discount modules, however, the issue for us is to rely on account_payment_order whereas this module allows reconciliation from the bank statements.

**Comment by sbidoul** [comment]
@grindtildeath account_cash_discount_base does not depend on account_payment_order. It provides the base data on invoices, and I think it should be possible to build the bank statement reconciliation part on top of it ? account_cash_discount_payment builds on it to facilitate supplier payment with cash discount.

**Comment by grindtildeath** [comment]
@sbidoul That's true but the design choices are quite different in terms of storing discount values. We could consider trying to merge modules together but IMO it would require changes on both sides (i.e. this module but also the other)

Then, there is also a third module doing the same thing in OCA but with a different scope: https://github.com/OCA/account-payment/tree/14.0/account_payment_term_discount
 
So I don't see why we shouldn't have the different alternatives available on existing versions. I guess we could try to converge all the modules together (maybe for v15.0), but it would require some time to do it properly so that it answers everyone's needs... Maybe we can start by opening an issue and try to involve all the contributors?

**Comment by sbidoul** [comment]
> So I don't see why we shouldn't have the different alternatives available on existing versions. 

Sure, I'm not blocking anything :)

> I guess we could try to converge all the modules together (maybe for v15.0), but it would require some time to do it properly so that it answers everyone's needs... Maybe we can start by opening an issue and try to involve all the contributors?

Good idea. Will you do it as it seems you have done some comparisons already ?

**Comment by grindtildeath** [comment]
FTR: https://github.com/OCA/account-payment/issues/462

File: account_financial_discount/models/account_reconcile_model.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        res = []
+        for line in move_lines:
+            tax_line = line.discount_tax_line_id
+            if not tax_line:
+                continue
+            tax_write_off_vals = {
+                "name": tax_line.name,
+                "account_id": tax_line.account_id.id,
+                "debit": tax_line.credit and line.amount_discount_tax or 0,
+                "credit": tax_line.debit and line.amount_discount_tax or 0,
```
Comment (grindtildeath) [inline account_financial_discount/models/account_reconcile_model.py:141]:
```suggestion
                "credit": tax_line.debit and -line.amount_discount_tax or 0,
```
While working on mig to v14.0 I'm wondering if we shouldn't have this here (for vendor bills).

File: account_financial_discount/models/account_reconcile_model.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                "reconcile_model_id": self.id,
+            }
+            # Deduce tax amount from fin. disc. write-off
+            if fin_disc_write_off_vals.get("credit"):
+                fin_disc_write_off_vals["credit"] -= line.amount_discount_tax
```
Comment (grindtildeath) [inline account_financial_discount/models/account_reconcile_model.py:146]:
```suggestion
                fin_disc_write_off_vals["credit"] += line.amount_discount_tax
```
While working on mig to v14.0 I'm wondering if we shouldn't have this here (for vendor bills)

**Pull request closed without merging**